istioctl/cmd/istioctl/istioctl_test.go

```diff
@@ -0,0 +1,582 @@
+			}
+		}
+	}
+}
+
+func TestDelete(t *testing.T) {
+	cases := []struct {
+		configs       []model.Config
+		args          []string
+		filename      string
+		wantOutput    string
+		wantError     *regexp.Regexp
+		wantException bool
+	}{
+		{ // case 0
+			[]model.Config{},
+			[]string{"routerule", "unknown"},
+			"",
+			"",
+			regexp.MustCompile("^$"),
+			true,
+		},
+		{ // case 1
+			testRouteRules,
+			[]string{"routerule", "a"},
+			"",
+			`Deleted config: routerule a
+`,
+			regexp.MustCompile("^$"),
+			false,
+		},
+		{ // case 2 - delete by filename of istio config which doesn't exist
+			testRouteRules,
+			[]string{},
+			"convert/testdata/v1alpha1/route-rule-80-20.yaml",
+			"",
+			regexp.MustCompile("^$"),
+			true,
+		},
+	}
+
+	for i, c := range cases {
+		// Override the client factory used by main.go
+		clientFactory = mockClientFactoryGenerator(c.configs)
+
+		// set the filename
+		file = c.filename
+
+		// run getCmd capturing output
+		stdOutput, errOutput, fErr := captureOutput(
+			func() error {
+				err := rootCmd.PersistentPreRunE(deleteCmd, c.args)
+				if err != nil {
+					return multierror.Prefix(err, "Could not prerun DELETE command")
+				}
+				err = deleteCmd.RunE(deleteCmd, c.args)
+				if err != nil {
+					return multierror.Prefix(err, "Could not run DELETE command")
+				}
+				return nil
+			})
+
+		if c.wantOutput != stdOutput {
+			t.Errorf("Stdout didn't match for delete case %d \"istioctl delete %v\": Expected %q, got %q.  Stderr was %q",
+				i, strings.Join(c.args, " "),
+				c.wantOutput, stdOutput, errOutput)
+		}
+
+		if !c.wantError.MatchString(errOutput) {
+			t.Errorf("Stderr didn't match for delete case %d %v: Expected %v, got %q", i, c.args, c.wantError, errOutput)
+		}
+
+		if c.wantException {
+			if fErr == nil {
+				t.Errorf("Wanted an exception for case %d %v, didn't get one, output was %q", i, c.args, stdOutput)
+			}
+		} else {
+			if fErr != nil {
+				t.Errorf("Unwanted exception for case %d %v: %v", i, c.args, fErr)
+			}
+		}
+	}
+}
+
+// mockClientFactoryGenerator creates a factory for model.ConfigStore preloaded with data
+func mockClientFactoryGenerator(configs []model.Config) func() (model.ConfigStore, error) {
+	outFactory := func() (model.ConfigStore, error) {
+		// Initialize the real client to get the supported config types
+		realClient, err := newClient()
+		if err != nil {
+			return nil, err
+		}
+
+		// Initialize memory based model.ConfigStore with configs
+		outConfig := memory.Make(realClient.ConfigDescriptor())
+		for _, config := range configs {
+			if _, err := outConfig.Create(config); err != nil {
+				return nil, err
+			}
+		}
+
+		// Wrap the memory ConfigStore so List() is sorted
+		return sortedConfigStore{store: outConfig}, nil
+	}
+
+	return outFactory
+}
+
+// captureOutput invokes f capturing the output sent to stderr and stdout
+func captureOutput(f func() error) (string, string, error) {
```

Comment:
Could we capture cobra output more directly using the `SetOutput` helper function?

```
var out bytes.Buffer
rootCmd.SetOutput(out)
```